File: server/app/live_instrument_matrix.py
"""Live instrument matrix — tools/live_instrument_matrix.json."""

from __future__ import annotations

import json
import re
from dataclasses import dataclass
from pathlib import Path
from typing import Any

from app.payload_optimization import (
    process_dynamic_vocal_and_instrument_payload,
    vocal_option_from_audio_environment,
)

_JSON_PATH = Path(__file__).resolve().parents[2] / "tools" / "live_instrument_matrix.json"
_DEFAULT_KEY = "default"
_LEGACY_ALIASES: dict[str, str] = {}


@dataclass(frozen=True)
class LiveInstrument:
    id: str
    name: str
    category: str
    default_articulation: str
    mix_role: str
    prompt_text: str = ""
    alias_hints: tuple[str, ...] = ()

    @property
    def prompt_label(self) -> str:
        p = (self.prompt_text or "").strip()
        return p if p else self.name


_LIVE_AMBIENCE_STUDIO = (
    "dead-room isolation",
    "close-mic studio capture",
    "dry acoustic room",
)


_MATRIX: dict[str, list[LiveInstrument]] | None = None
_SCHEMA: dict[str, Any] | None = None


def _parse_json() -> dict[str, Any]:
    raw = json.loads(_JSON_PATH.read_text(encoding="utf-8"))
    if "genres" in raw:
        return {
            "schema_version": str(raw.get("schemaVersion", "1.1.0")),
            "category_taxonomy": dict(raw.get("categoryTaxonomy", {})),
            "category_classification_rules": list(
                raw.get("categoryClassificationRules", [])
            ),
            "genre_preset_alias": dict(raw.get("genrePresetAlias", {})),
            "genres": dict(raw["genres"]),
        }
    return {
        "schema_version": "1.0.0",
        "category_taxonomy": {},
        "category_classification_rules": [],
        "genre_preset_alias": {},
        "genres": raw,
    }


def _raw_json() -> dict[str, Any]:
    global _SCHEMA  # noqa: PLW0603
    if _SCHEMA is None:
        _SCHEMA = _parse_json()
    return _SCHEMA


def _row_to_instrument(row: dict[str, Any]) -> LiveInstrument:
    alias_raw = row.get("aliases") or row.get("aliasHints") or []
    if isinstance(alias_raw, str):
        aliases = tuple(a.strip() for a in alias_raw.split("|") if a.strip())
    else:
        aliases = tuple(str(a).strip() for a in alias_raw if str(a).strip())
    return LiveInstrument(
        id=str(row["id"]),
        name=str(row["name"]),
        category=str(row.get("category", "")),
        default_articulation=str(row.get("defaultArticulation", "")),
        mix_role=str(row.get("mixRole", "")),
        prompt_text=str(row.get("promptText") or ""),
        alias_hints=aliases,
    )


def selection_implies_live_ambience(real_instrumentals: str) -> bool:
    return "live" in str(real_instrumentals or "").lower()


def augment_avoid_clause(*, avoid: str, real_instrumentals: str) -> str:
    if not selection_implies_live_ambience(real_instrumentals):
        return str(avoid or "").strip()
    seen: set[str] = set()
    out: list[str] = []
    for clause in [*( [avoid.strip()] if avoid and avoid.strip() else []), *_LIVE_AMBIENCE_STUDIO]:
        key = clause.lower()
        if key not in seen:
            seen.add(key)
            out.append(clause)
    return ", ".join(out)


def _all_instruments_union() -> list[LiveInstrument]:
    seen: set[str] = set()
    out: list[LiveInstrument] = []
    for rows in _raw_json()["genres"].values():
        for row in rows:
            inst = _row_to_instrument(dict(row))
            if inst.id not in seen:
                seen.add(inst.id)
                out.append(inst)
    return out


def _load() -> dict[str, list[LiveInstrument]]:
    global _MATRIX  # noqa: PLW0603
    if _MATRIX is not None:
        return _MATRIX
    schema = _raw_json()
    out: dict[str, list[LiveInstrument]] = {}
    for genre, rows in schema["genres"].items():
        out[genre] = [_row_to_instrument(dict(row)) for row in rows]
    _MATRIX = out
    return out


def invalidate_matrix_cache() -> None:
    """Clear cached matrix after JSON regeneration."""
    global _MATRIX, _SCHEMA  # noqa: PLW0603
    _MATRIX = None
    _SCHEMA = None


def bundle_key_for_genre(genre: str | None) -> str | None:
    """Exact alias match (case-insensitive), then longest substring match."""
    trimmed = (genre or "").strip()
    if not trimmed:
        return None
    aliases = _raw_json().get("genre_preset_alias", {})
    keys = set(_load().keys())
    for alias, target in aliases.items():
        if alias.lower() == trimmed.lower() and target in keys:
            return str(target)
    best = ""
    lower = trimmed.lower()
    for alias, target in aliases.items():
        alias_lower = alias.lower()
        if alias_lower in lower and target in keys and len(alias_lower) > len(best):
            best = str(target)
    return best or None


def voices_for_genre(genre: str | None) -> list[dict[str, Any]]:
    key = bundle_key_for_genre(genre)
    if not key:
        return []
    return [dict(row) for row in _raw_json()["genres"].get(key, [])]


def _bundle_for_label(label: str) -> str | None:
    return bundle_key_for_genre(label)


def _legacy_hit(text: str, keys: set[str]) -> str | None:
    lower = text.lower()
    for alias, target in _LEGACY_ALIASES.items():
        if alias in lower and target in keys:
            return target
    return None


def resolve_genre_instrument_key(primary: str, fusion: str = "") -> str:
    matrix = _load()
    keys = set(matrix.keys())
    blob = f"{primary} {fusion}".lower().strip()

    if fusion.strip():
        primary_hit = _bundle_for_label(primary) or _legacy_hit(primary, keys)
        if primary_hit and primary_hit in keys:
            return primary_hit

    for label in (primary, fusion):
        hit = _bundle_for_label(label)
        if hit and hit in keys:
            return hit

    legacy = _legacy_hit(blob, keys)
    if legacy and legacy in keys:
        return legacy

    partial = bundle_key_for_genre(blob)
    if partial and partial in keys:
        return partial

    return _best_genre_key_from_blob(blob, keys) or _DEFAULT_KEY


def _best_genre_key_from_blob(blob: str, keys: set[str]) -> str:
    best = ""
    for key in keys:
        phrase = key.replace("_", " ")
        if phrase in blob and len(phrase) > len(best):
            best = key
    return best


def instruments_for_genre(
    primary: str,
    fusion: str = "",
    *,
    vocal_option_ui: str = "close-mic",
) -> list[LiveInstrument]:
    key = resolve_genre_instrument_key(primary, fusion)
    genres = _raw_json()["genres"]
    raw_rows = [dict(r) for r in genres.get(key, genres[_DEFAULT_KEY])]
    patched = process_dynamic_vocal_and_instrument_payload(raw_rows, vocal_option_ui)
    return [_row_to_instrument(row) for row in patched]


def parse_instrument_selection(raw: str) -> list[str]:
    return [t.strip() for t in raw.split(",") if t.strip()]


def _power_codes_active(power_codes: str) -> bool:
    return "/L99" in (power_codes or "").upper()


def _gear_modifier(inst: LiveInstrument, l99: bool) -> str:
    if not l99:
        return ""
    if inst.category == "guitar":
        return " (vintage Fender amp, tube warmth)"
    if inst.category == "keys":
        return " (Neve 1073 preamp, analog warmth)"
    if inst.category in ("horns", "strings", "saxophones", "trumpets", "winds"):
        return " (Neumann U47 close-mic'd, dry room)"
    if inst.category in ("bass", "melodic_bass"):
        return " (Ampeg SVT warmth, tight DI blend)"
    return ""


def _token_matches_instrument(token: str, inst: LiveInstrument) -> bool:
    t = re.sub(r"\s+", " ", token.lower().strip())
    if not t:
        return False
    if inst.id.lower() == t or inst.name.lower() == t:
        return True
    label = inst.prompt_label.lower()
    if label == t or t in label or label in t:
        return True
    for alias in inst.alias_hints:
        a = alias.lower()
        if a == t or t in a or a in t:
            return True
    return False


def _match_selected(
    available: list[LiveInstrument],
    tokens: list[str],
    *,
    fallback_catalog: list[LiveInstrument] | None = None,
) -> list[LiveInstrument]:
    out: list[LiveInstrument] = []
    seen: set[str] = set()
    catalog = fallback_catalog if fallback_catalog is not None else available

    for token in tokens:
        hit: LiveInstrument | None = None
        for inst in available:
            if _token_matches_instrument(token, inst):
                hit = inst
                break
        if hit is None:
            for inst in catalog:
                if _token_matches_instrument(token, inst):
                    hit = inst
                    break
        if hit and hit.id not in seen:
            seen.add(hit.id)
            out.append(hit)
    return out


def _normalize_version(version: str | None) -> str:
    from app.suno_version import PREFERRED, density_key_for

    key = density_key_for(version or PREFERRED)
    if key == "v4.5":
        return "v4.5"
    if key == "v5.5":
        return "v5.5pro"
    return "v5"


def generate_live_instrument_prompt(
    genre: str,
    selection_raw: str,
    version: str | None = None,
    power_codes: str = "",
    fusion: str = "",
    *,
    vocal_option_ui: str = "close-mic",
) -> tuple[str, str]:
    """Returns (style_injection, meta_tag_injection)."""
    tokens = parse_instrument_selection(selection_raw)
    if not tokens:
        return "", ""

    key = resolve_genre_instrument_key(genre, fusion)
    available = instruments_for_genre(genre, fusion, vocal_option_ui=vocal_option_ui)
    catalog = _all_instruments_union()
    selected = _match_selected(available, tokens, fallback_catalog=catalog)
    if not selected:
        return "", ""

    v = _normalize_version(version)
    l99 = _power_codes_active(power_codes)

    descriptions: list[str] = []
    for inst in selected:
        label = inst.prompt_label
        mod = _gear_modifier(inst, l99)
        use_prompt_only = bool((inst.prompt_text or "").strip())
        if v == "v4.5":
            if use_prompt_only:
                descriptions.append(f"{label}, {inst.mix_role}{mod}")
            else:
                descriptions.append(
                    f"{label}, {inst.default_articulation}, {inst.mix_role}{mod}"
                )
        elif v == "v5":
            if use_prompt_only:
                descriptions.append(
                    f"featuring {label} sitting in the {inst.mix_role}{mod}"
                )
            else:
                descriptions.append(
                    f"featuring {inst.default_articulation} {label} "
                    f"sitting in the {inst.mix_role}{mod}"
                )
        else:
            if use_prompt_only:
                descriptions.append(
                    f"driven by {label}{mod}, perfectly seated in the {inst.mix_role}"
                )
            else:
                descriptions.append(
                    f"driven by a {inst.default_articulation} {label}{mod}, "
                    f"perfectly seated in the {inst.mix_role}"
                )

    if v == "v4.5":
        style = f", {', '.join(descriptions)}"
        meta = f"[{selected[0].prompt_label} Feature]"
    elif v == "v5":
        style = f". {', '.join(descriptions)}."
        meta = f"[Instrumental: {' and '.join(s.prompt_label for s in selected)} interplay]"
    else:
        style = f". The arrangement is elevated by {', and '.join(descriptions)}."
        primary = selected[0]
        meta = (
            f"[Instrumental Break: Feature {primary.default_articulation} "
            f"{primary.prompt_label}, {primary.mix_role}, dynamic lift, subtle tape saturation]"
        )
        if len(selected) > 1:
            secondary = selected[1]
            meta += (
                f"\n[Bridge: Intimate interplay between {secondary.prompt_label} "
                f"and vocals, {secondary.mix_role}]"
            )

    return style, meta


def live_instrument_user_block(
    *,
    primary_genre: str,
    sub_genre_fusion: str = "",
    selection_raw: str,
    suno_version: str | None = None,
    power_codes: str = "",
    audio_environment_mode: str = "studio_isolated",
) -> str:
    raw = (selection_raw or "").strip()
    if not raw:
        return ""

    vocal_option = vocal_option_from_audio_environment(audio_environment_mode)
    key = resolve_genre_instrument_key(primary_genre, sub_genre_fusion)
    style_inj, meta_inj = generate_live_instrument_prompt(
        primary_genre,
        raw,
        version=suno_version,
        power_codes=power_codes,
        fusion=sub_genre_fusion,
        vocal_option_ui=vocal_option,
    )
    matched = _match_selected(
        instruments_for_genre(
            primary_genre,
            sub_genre_fusion,
            vocal_option_ui=vocal_option,
        ),
        parse_instrument_selection(raw),
        fallback_catalog=_all_instruments_union(),
    )

    lines = [
        f"Matrix schema: {_raw_json().get('schema_version', '1.0.0')}",
        f"Matched genre profile: [{key}]",
        f"Vocal backing patch: [{vocal_option}] (from audio_environment_mode={audio_environment_mode or 'studio_isolated'})",
        f"User selection: {raw}",
    ]
    if matched:
        for inst in matched:
            lines.append(
                f"• {inst.prompt_label} ({inst.name}): {inst.default_articulation} | mix: {inst.mix_role}"
            )
    else:
        lines.append(
            "• Free-text instruments — describe articulation + mix placement per protocol; "
            "frame atypical pairings as intentional fusion."
        )
    if style_inj:
        lines.append(f"Block 1 styleInjection (weave into producer prose): {style_inj}")
    if meta_inj:
        lines.append(f"Block 2 metaTagInjection (place in timeline): {meta_inj}")
    lines.append(
        "Never list bare instrument names. No artist names. Honor version-aware injection from REAL INSTRUMENT PROTOCOL."
    )
    body = "\n".join(lines)
    return (
        "[REAL INSTRUMENT ACCOMPANIMENT] (MANDATORY: Feature the following "
        "instruments prominently. Emphasize their natural, acoustic character and "
        "the specified articulations. This is a production requirement, not a "
        f"suggestion.):\n{body}"
    )

File: lib/core/utils/live_instrument_matrix.dart
import 'live_instrument_matrix_data.dart';

/// Genre-accurate live instrument profiles (tools/live_instrument_matrix.json).
class LiveInstrument {
  const LiveInstrument({
    required this.id,
    required this.name,
    required this.category,
    required this.defaultArticulation,
    required this.mixRole,
    this.promptText = '',
    this.aliasHints = const [],
  });

  final String id;
  final String name;
  final String category;
  final String defaultArticulation;
  final String mixRole;

  /// LLM-facing label when set; UI keeps [name].
  final String promptText;
  final List<String> aliasHints;

  String get promptLabel {
    final p = promptText.trim();
    return p.isNotEmpty ? p : name;
  }
}

class LiveInstrumentMatrix {
  LiveInstrumentMatrix._();

  static const _defaultKey = 'default';

  static const _liveAmbienceStudioClauses = [
    'dead-room isolation',
    'close-mic studio capture',
    'dry acoustic room',
  ];

  static LiveInstrument _fromRow(Map<String, String> row) {
    final aliasRaw = row['aliasHints'] ?? '';
    return LiveInstrument(
      id: row['id']!,
      name: row['name']!,
      category: row['category']!,
      defaultArticulation: row['defaultArticulation']!,
      mixRole: row['mixRole']!,
      promptText: row['promptText'] ?? '',
      aliasHints: aliasRaw.isEmpty
          ? const []
          : aliasRaw.split('|').map((s) => s.trim()).where((s) => s.isNotEmpty).toList(),
    );
  }

  static const _legacyAliases = <String, String>{};

  /// When selection text includes "Live", reinforce studio capture (positive tokens only).
  static bool selectionImpliesLiveAmbience(String realInstrumentals) =>
      realInstrumentals.toLowerCase().contains('live');

  static String augmentAvoidClause({
    required String avoid,
    required String realInstrumentals,
  }) {
    if (!selectionImpliesLiveAmbience(realInstrumentals)) {
      return avoid.trim();
    }
    final seen = <String>{};
    final out = <String>[];
    for (final clause in [
      if (avoid.trim().isNotEmpty) avoid.trim(),
      ..._liveAmbienceStudioClauses,
    ]) {
      final key = clause.toLowerCase();
      if (seen.add(key)) out.add(clause);
    }
    return out.join(', ');
  }

  static String? _bundleForLabel(String label) {
    final key = LiveInstrumentMatrixData.bundleKeyForGenre(label);
    if (key == null || !LiveInstrumentMatrixData.byGenre.containsKey(key)) {
      return null;
    }
    return key;
  }

  static String _bestPartialAlias(String blob) {
    final key = LiveInstrumentMatrixData.bundleKeyForGenre(blob);
    if (key != null && LiveInstrumentMatrixData.byGenre.containsKey(key)) {
      return key;
    }
    return '';
  }

  static String? _legacyHit(String text, Set<String> keys) {
    final lower = text.toLowerCase();
    for (final entry in _legacyAliases.entries) {
      if (lower.contains(entry.key) && keys.contains(entry.value)) {
        return entry.value;
      }
    }
    return null;
  }

  static String resolveGenreKey(String primary, String fusion) {
    final keys = LiveInstrumentMatrixData.byGenre.keys.toSet();
    final blob = '${primary.trim()} ${fusion.trim()}'.toLowerCase();

    if (fusion.trim().isNotEmpty) {
      final primaryHit = _bundleForLabel(primary) ?? _legacyHit(primary, keys);
      if (primaryHit != null) return primaryHit;
    }

    for (final label in [primary, fusion]) {
      final hit = _bundleForLabel(label);
      if (hit != null) return hit;
    }

    final legacy = _legacyHit(blob, keys);
    if (legacy != null) return legacy;

    final partial = _bestPartialAlias(blob);
    if (partial.isNotEmpty) return partial;

    return _bestGenreKeyFromBlob(blob).ifEmpty(_defaultKey);
  }

  static String _bestGenreKeyFromBlob(String blob) {
    var best = '';
    for (final key in LiveInstrumentMatrixData.byGenre.keys) {
      final phrase = key.replaceAll('_', ' ');
      if (blob.contains(phrase) && phrase.length > best.length) {
        best = key;
      }
    }
    return best;
  }

  static List<LiveInstrument> instrumentsForGenre(
    String primary,
    String fusion,
  ) {
    final key = resolveGenreKey(primary, fusion);
    final rows = LiveInstrumentMatrixData.byGenre[key] ??
        LiveInstrumentMatrixData.byGenre[_defaultKey]!;
    return rows.map(_fromRow).toList();
  }

  static List<LiveInstrument> _allInstrumentsUnion() {
    final seen = <String>{};
    final out = <LiveInstrument>[];
    for (final rows in LiveInstrumentMatrixData.byGenre.values) {
      for (final row in rows) {
        final inst = _fromRow(row);
        if (seen.add(inst.id)) out.add(inst);
      }
    }
    return out;
  }

  static List<String> parseSelection(String raw) =>
      raw.split(',').map((t) => t.trim()).where((t) => t.isNotEmpty).toList();

  static bool _tokenMatchesInstrument(String token, LiveInstrument inst) {
    final t = token.toLowerCase().trim();
    if (t.isEmpty) return false;
    if (inst.id.toLowerCase() == t || inst.name.toLowerCase() == t) {
      return true;
    }
    final label = inst.promptLabel.toLowerCase();
    if (label == t || t.contains(label) || label.contains(t)) {
      return true;
    }
    for (final alias in inst.aliasHints) {
      final a = alias.toLowerCase();
      if (a == t || t.contains(a) || a.contains(t)) return true;
    }
    return false;
  }

  static List<LiveInstrument> _matchSelected(
    List<LiveInstrument> available,
    List<String> tokens, {
    List<LiveInstrument>? fallbackCatalog,
  }) {
    final out = <LiveInstrument>[];
    final catalog = fallbackCatalog ?? available;
    for (final token in tokens) {
      LiveInstrument? hit;
      for (final inst in available) {
        if (_tokenMatchesInstrument(token, inst)) {
          hit = inst;
          break;
        }
      }
      if (hit == null) {
        for (final inst in catalog) {
          if (_tokenMatchesInstrument(token, inst)) {
            hit = inst;
            break;
          }
        }
      }
      if (hit != null && !out.any((x) => x.id == hit!.id)) out.add(hit);
    }
    return out;
  }

  static String _gearModifier(LiveInstrument inst, bool l99) {
    if (!l99) return '';
    switch (inst.category) {
      case 'guitar':
        return ' (vintage Fender amp, tube warmth)';
      case 'keys':
        return ' (Neve 1073 preamp, analog warmth)';
      case 'horns':
      case 'strings':
      case 'saxophones':
      case 'trumpets':
      case 'winds':
        return " (Neumann U47 close-mic'd, dry room)";
      case 'bass':
      case 'melodic_bass':
        return ' (Ampeg SVT warmth, tight DI blend)';
      default:
        return '';
    }
  }

  static String _normalizeVersion(String version) {
    final v = version.trim().toLowerCase();
    if (v == 'v4.5') return 'v4.5';
    if (v.startsWith('v5.5')) return 'v5.5pro';
    return 'v5';
  }

  static ({String styleInjection, String metaTagInjection}) generatePrompt({
    required String genre,
    required String selectionRaw,
    required String sunoVersion,
    String fusionGenre = '',
    String powerCodes = '',
  }) {
    final tokens = parseSelection(selectionRaw);
    if (tokens.isEmpty) {
      return (styleInjection: '', metaTagInjection: '');
    }
    final available = instrumentsForGenre(genre, fusionGenre);
    final catalog = _allInstrumentsUnion();
    final selected = _matchSelected(
      available,
      tokens,
      fallbackCatalog: catalog,
    );
    if (selected.isEmpty) {
      return (styleInjection: '', metaTagInjection: '');
    }

    final v = _normalizeVersion(sunoVersion);
    final l99 = powerCodes.toUpperCase().contains('/L99');

    final descriptions = selected.map((inst) {
      final label = inst.promptLabel;
      final mod = _gearModifier(inst, l99);
      final usePromptOnly = inst.promptText.trim().isNotEmpty;
      if (v == 'v4.5') {
        if (usePromptOnly) {
          return '$label, ${inst.mixRole}$mod';
        }
        return '$label, ${inst.defaultArticulation}, ${inst.mixRole}$mod';
      }
      if (v == 'v5') {
        if (usePromptOnly) {
          return 'featuring $label sitting in the ${inst.mixRole}$mod';
        }
        return 'featuring ${inst.defaultArticulation} $label sitting in the ${inst.mixRole}$mod';
      }
      if (usePromptOnly) {
        return 'driven by $label$mod, perfectly seated in the ${inst.mixRole}';
      }
      return 'driven by a ${inst.defaultArticulation} $label$mod, perfectly seated in the ${inst.mixRole}';
    }).toList();

    if (v == 'v4.5') {
      return (
        styleInjection: ', ${descriptions.join(', ')}',
        metaTagInjection: '[${selected.first.promptLabel} Feature]',
      );
    }
    if (v == 'v5') {
      return (
        styleInjection: '. ${descriptions.join(', ')}.',
        metaTagInjection:
            '[Instrumental: ${selected.map((s) => s.promptLabel).join(' and ')} interplay]',
      );
    }
    var meta =
        '[Instrumental Break: Feature ${selected.first.defaultArticulation} ${selected.first.promptLabel}, ${selected.first.mixRole}, dynamic lift, subtle tape saturation]';
    if (selected.length > 1) {
      final sec = selected[1];
      meta +=
          '\n[Bridge: Intimate interplay between ${sec.promptLabel} and vocals, ${sec.mixRole}]';
    }
    return (
      styleInjection:
          '. The arrangement is elevated by ${descriptions.join(', and ')}.',
      metaTagInjection: meta,
    );
  }

  static String userBlockDirective({
    required String primaryGenre,
    String subGenreFusion = '',
    required String selectionRaw,
    required String sunoVersion,
    String powerCodes = '',
  }) {
    final raw = selectionRaw.trim();
    if (raw.isEmpty) return '';

    final key = resolveGenreKey(primaryGenre, subGenreFusion);
    final catalog = _allInstrumentsUnion();
    final matched = _matchSelected(
      instrumentsForGenre(primaryGenre, subGenreFusion),
      parseSelection(raw),
      fallbackCatalog: catalog,
    );
    final prompt = generatePrompt(
      genre: primaryGenre,
      selectionRaw: raw,
      sunoVersion: sunoVersion,
      fusionGenre: subGenreFusion,
      powerCodes: powerCodes,
    );

    final lines = <String>[
      'Matrix schema: ${LiveInstrumentMatrixData.schemaVersion}',
      'Matched genre profile: [$key]',
      'User selection: $raw',
    ];
    if (matched.isNotEmpty) {
      for (final inst in matched) {
        final label = inst.promptLabel;
        lines.add(
          '• $label (${inst.name}): ${inst.defaultArticulation} | mix: ${inst.mixRole}',
        );
      }
    } else {
      lines.add(
        '• Free-text instruments — describe articulation + mix placement per protocol; '
        'frame atypical pairings as intentional fusion.',
      );
    }
    if (prompt.styleInjection.isNotEmpty) {
      lines.add(
        'Block 1 styleInjection (weave into producer prose): ${prompt.styleInjection}',
      );
    }
    if (prompt.metaTagInjection.isNotEmpty) {
      lines.add(
        'Block 2 metaTagInjection (place in timeline): ${prompt.metaTagInjection}',
      );
    }
    lines.add(
      'Never list bare instrument names. No artist names. Honor version-aware injection from REAL INSTRUMENT PROTOCOL.',
    );

    final body = lines.join('\n');
    return '[REAL INSTRUMENT ACCOMPANIMENT] (MANDATORY: Feature the following '
        'instruments prominently. Emphasize their natural, acoustic character and '
        'the specified articulations. This is a production requirement, not a '
        'suggestion.):\n$body';
  }
}

extension _IfEmpty on String {
  String ifEmpty(String fallback) => isEmpty ? fallback : this;
}
